### PR TITLE
Monitor (& log) KV retries during Graveler operations

### LINKED ISSUE #4563: Monitor branch operation retries
This issue is asking for metrics that will monitor retry operations during
mergeIntoBranch and other operations.

The upcoming LakeFSOutputCommitter performs numerous API branch operations
concurrently, especially mergeIntoBranch.  These can fail, requiring retries
-- this expected.  Because this an automatic (Spark) process performs the
operations, percolating these failure upwards is extremely expensive, and
users will prefer to retry over much longer periods of time than are
reasonable for an API call.  These metrics will allow us and users to
estimate _what parameters to use_, and to _measure performance of parameters
in use_.

### PR DESCRIPTION
**Pull request by arielshaqed**
This will show up when branch operations or other updates collide.

Fixes #4563.

### CODE REVIEW

**Comment by arielshaqed** [comment]
I will pull with either reviewer's approval.  Invited @Isan-Rivkin so he can
see deep lakeFS fun.

File: pkg/graveler/graveler.go [hunk 1]
```diff
@@ -19,6 +21,14 @@ import (
 	"google.golang.org/protobuf/types/known/timestamppb"
 )
 
+var kvRetriesCounter = promauto.NewCounterVec(
+	prometheus.CounterOpts{
+		Name: "graveler_kv_retries",
+		Help: "Number of retries performed by Graveler to update value.",
+	},
+	[]string{"operation", "repository"},
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:29]: We shouldn't collect repos as metrics for 2 reasons:
1. Possible high cardinality
2. Possible user sensitive info. Again, not very likely but still.

It's not that any of the above is a high risk, but the gain of collecting the repos isn't big either. We don't collect repos anywhere else too.

File: pkg/graveler/graveler.go [hunk 2]
```diff
@@ -1217,11 +1229,28 @@ func (g *Graveler) UpdateBranch(ctx context.Context, repository *RepositoryRecor
...
+		"retries":    retries,
+		"repository": repositoryID,
+		"branch":     branchID,
+		"operation":  operation,
+	})
+	if retries > 0 {
+		l.Info("Operation retried")
+	} else {
+		l.Trace("No retries needed")
+	}
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:1246]: Can we have just one place where we log retries? You have Info + Trace here and Debug below. I think `Info` is too verbose for a somewhat expected retry, but anyway let's have it there or here and not both.

File: pkg/graveler/graveler.go [hunk 3]
```diff
@@ -1655,10 +1691,12 @@ func (g *Graveler) Delete(ctx context.Context, repository *RepositoryRecord, bra
 	}
 
 	log := g.log(ctx).WithFields(logging.Fields{"key": key, "operation": "delete"})
-	return g.safeBranchWrite(ctx, log, repository, branchID,
+	retries, err := g.safeBranchWrite(ctx, log, repository, branchID,
 		safeBranchWriteOptions{}, func(branch *Branch) error {
 			return g.deleteUnsafe(ctx, repository, branch, key, nil)
 		})
+	g.monitorRetries(ctx, retries, repository.RepositoryID, branchID, "delete")
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:1698]: I do feel like this pattern should become the function itself or at least we should wrap this common functionality with another func:
```suggestion
	err := g.safeBranchWriteWithMonitor(ctx, log, repository, branchID,
		safeBranchWriteOptions{}, func(branch *Branch) error {
			return g.deleteUnsafe(ctx, repository, branch, key, nil)
		}
```

or just 
```suggestion
          err := g.safeBranchWrite(ctx, log, repository, branchID,
		safeBranchWriteOptions{}, func(branch *Branch) error {
			return g.deleteUnsafe(ctx, repository, branch, key, nil)
		})
```

File: pkg/graveler/graveler.go [hunk 4]
```diff
@@ -2283,7 +2328,9 @@ func (g *Graveler) Revert(ctx context.Context, repository *RepositoryRecord, bra
 		parentNumber--
 	}
 
-	if err := g.prepareForCommitIDUpdate(ctx, repository, branchID); err != nil {
+	retries, err := g.prepareForCommitIDUpdate(ctx, repository, branchID)
+	g.monitorRetries(ctx, retries, repository.RepositoryID, branchID, "revert")
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:2332]: Same here

File: pkg/graveler/graveler.go (hunk 3 above)
Comment (itaiad200) [reply pkg/graveler/graveler.go:1698]: Mostly worried about adding `safeBranchWrite` in the future without monitoring the retries.

**Review by itaiad200 — CHANGES_REQUESTED** [review]
Thanks! Everything is easily fixable, let me know if you have any questions

**Review by N-o-Z — COMMENTED** [review]
We don't have a general retry operation context, which causes us to have several replicated retry logic as @itaiad200 stated. We can easily miss logging retry operations this way, one example is retryRepoMetadataUpdate. I understand that this PR focuses mainly on the branch update retry mechanism, but since the title of the PR says "monitor retries during Graveler operations" I thought this is worth a mention

**Review by arielshaqed — COMMENTED** [review]
Thanks; PTAL!

**Review by itaiad200 — APPROVED** [review]
LGTM

**Comment by arielshaqed** [comment]
Thanks!

Pulling as soon as GitHub actions notices that GitHub actions have succeeded.

**Comment by arielshaqed** [comment]
> We don't have a general retry operation function, which causes us to have several replicated retry logics as @itaiad200 stated. We can easily miss logging retry operations this way, one example is retryRepoMetadataUpdate. I understand that this PR focuses mainly on the branch update retry mechanism, but since the title of the PR says "monitor retries during Graveler operations" I thought this is worth a mention

Unfortunately each retry is subtly different.  We should probably collect more context about what we _want_ each to do and then refactor.  (For instance, we don't always back off!)

**Pull request merged**